### PR TITLE
{2023.06}[2023a,zen4] LAMMPS 2Aug2023 for zen4

### PR DESCRIPTION
**Pull request by laraPPr**
Building LAMMPS 2Aug2023 requires https://github.com/easybuilders/easybuild-easyblocks/pull/3336/files which is part of EasyBuild 4.9.3

### CODE REVIEW

**Comment by laraPPr** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

**Comment by laraPPr** [comment]
@boegel You can also try to build for A64FX has the same amount of missing dependencies as the zen4 stack

**Comment by boegel** [comment]
> @boegel You can also try to build for A64FX has the same amount of missing dependencies as the zen4 stack

I can, but if I do this for this PR, it will first try to install everything else already included in `easystacks/software.eessi.io/2023.06/zen4/*.yml`, so we should open a dedicated PR for `easystacks/software.eessi.io/2023.06/a64fx/*.yml`

**Comment by laraPPr** [comment]
> > @boegel You can also try to build for A64FX has the same amount of missing dependencies as the zen4 stack
> 
> I can, but if I do this for this PR, it will first try to install everything else already included in `easystacks/software.eessi.io/2023.06/zen4/*.yml`, so we should open a dedicated PR for `easystacks/software.eessi.io/2023.06/a64fx/*.yml`

Ah I'll open a seperate pr for `easystacks/software.eessi.io/2023.06/a64fx/*.yml`

**Comment by laraPPr** [comment]
since I started a new stack for EasyBuild 4.9.3 it will only pick up the changes in that EasyStack right. So in this case it would only pick up a64fx. But for clarity it might be best that we keep them seperate to some extend.

**Comment by laraPPr** [comment]
```
== 2024-09-17 12:01:03,066 run.py:689 DEBUG cmd "python -c 'from archspec.cpu import host; print(host())'" exited with exit code 0 and output:

x86_64_v4
```

The output should be zen4 not sure why this happened this did not happen when building at Gent with `zen4`

**Comment by boegel** [comment]
> since I started a new stack for EasyBuild 4.9.3 it will only pick up the changes in that EasyStack right. So in this case it would only pick up a64fx. But for clarity it might be best that we keep them seperate to some extend.

That's true, I overlooked that.

**Comment by boegel** [comment]
> ```
> == 2024-09-17 12:01:03,066 run.py:689 DEBUG cmd "python -c 'from archspec.cpu import host; print(host())'" exited with exit code 0 and output:
> 
> x86_64_v4
> ```
> 
> The output should be zen4 not sure why this happened this did not happen when building at Gent with `zen4`

This probably indicates that one of the CPU features expected by archspec for zen4 can't be found, so it "falls back" to `x86_64_v4`; see also https://github.com/archspec/archspec-json/blob/master/cpu/microarchitectures.json#L2140

It isn't the first time this happens in AWS or Azure, see for example https://github.com/archspec/archspec-json/issues/38

**Comment by laraPPr** [comment]
So than set it in a hook or add this to the easyBlock?

**Comment by ocaisa** [comment]
In this case it would need to be a hook, perhaps we should be doing this anyway given that we use archdetect rather than archspec.

You can use a parse hook that sets the correct value for `kokkos_arch`, with a table that maps between the environment variable `EESSI_SOFTWARE_SUBDIR` and the value Kokkos expects

**Comment by laraPPr** [comment]
Yeah we did the same in the pilot so I can copy and update that

**Comment by laraPPr** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

File: eb_hooks.py [hunk 1]
```diff
@@ -514,6 +516,23 @@ def pre_configure_hook_wrf_aarch64(self, *args, **kwargs):
...
+    - set kokkos_arch on x86_64/amd/zen4
+    """
+
+    cpu_target = get_eessi_envvar('EESSI_SOFTWARE_SUBDIR')
+    if self.name == 'LAMMPS':
+        if self.version == '2Aug2023':
+            if get_cpu_architecture() == X86_64:
+                if cpu_target == CPU_TARGET_ZEN4:
+                    # There is no support for ZEN4 in LAMMPS yet so falling back to ZEN3
+                    self.cfg['kokkos'] = 'ZEN3'
```
Comment (laraPPr) [inline eb_hooks.py:531]:
```suggestion
                    self.cfg['kokkos_arch'] = 'ZEN3'
```

**Comment by laraPPr** [comment]
Made a stupid mistake should be fixed now

**Comment by laraPPr** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

File: eb_hooks.py [hunk 2]
```diff
@@ -514,6 +516,23 @@ def pre_configure_hook_wrf_aarch64(self, *args, **kwargs):
...
 
+def pre_configure_hook_LAMMPS_zen4(self, *args, **kwargs):
+    """
+    pre-configure hook for LAMMPS:
+    - set kokkos_arch on x86_64/amd/zen4
+    """
+
+    cpu_target = get_eessi_envvar('EESSI_SOFTWARE_SUBDIR')
+    if self.name == 'LAMMPS':
+        if self.version == '2Aug2023':
```
Comment (laraPPr) [inline eb_hooks.py:527]:
```suggestion
        if self.version == '2Aug2023_update2':
```

**Comment by laraPPr** [comment]
bot: build repo:eessi.io-2023.06-software arch:zen4

**Comment by laraPPr** [comment]
Succes!!

**Review by boegel — APPROVED** [review]
lgtm

**Comment by boegel** [comment]
Ah snap, I merged before doing the deploy 🤦 

Should still work though, deploy triggered...

**Pull request merged**